File: ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/TransactionState.java
package net.corda.v5.ledger.utxo;

import net.corda.v5.base.annotations.CordaSerializable;
import net.corda.v5.base.annotations.DoNotImplement;
import net.corda.v5.base.types.MemberX500Name;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.security.PublicKey;

/**
 * Defines a transaction state, composed of a {@link ContractState} and associated transaction state information.
 *
 * @param <T> The underlying type of the {@link ContractState} instance.
 */
@DoNotImplement
@CordaSerializable
public interface TransactionState<T extends ContractState> {

    /**
     * Gets the {@link ContractState} of the current {@link TransactionState} instance.
     *
     * @return Returns the {@link ContractState} of the current {@link TransactionState} instance.
     */
    @NotNull
    T getContractState();

    /**
     * Gets the {@link ContractState} type of the current {@link TransactionState} instance.
     *
     * @return Returns the {@link ContractState} type of the current {@link TransactionState} instance.
     */
    @NotNull
    Class<T> getContractStateType();

    /**
     * Gets the {@link Contract} type of the current {@link TransactionState} instance.
     *
     * @return Returns the {@link Contract} type of the current {@link TransactionState} instance.
     */
    @NotNull
    Class<? extends Contract> getContractType();

    /**
     * Gets the notary name of the current {@link TransactionState} instance.
     *
     * @return Returns the notary of the current {@link TransactionState} instance.
     */
    @NotNull
    MemberX500Name getNotaryName();

    /**
     * Gets the notary key of the current {@link TransactionState} instance.
     *
     * @return Returns the notary of the current {@link TransactionState} instance.
     */
    @NotNull
    PublicKey getNotaryKey();


    /**
     * Gets the encumbrance of the current {@link TransactionState} instance.
     *
     * @return Returns the encumbrance of the current {@link TransactionState} instance.
     */
    @Nullable
    EncumbranceGroup getEncumbranceGroup();
}

File: ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionState.kt
package net.corda.v5.ledger.utxo

import net.corda.v5.base.annotations.CordaSerializable
import java.security.PublicKey

/**
 * Represents a transaction state, composed of a [ContractState] and associated [TransactionStateInformation].
 *
 * @constructor Creates a new instance of the [TransactionState] data class.
 * @param T The underlying type of the [ContractState] instance.
 * @property contractState The [ContractState] component of the current [TransactionState] instance.
 * @property information The [TransactionStateInformation] component of the current [TransactionState] instance.
 */
@CordaSerializable
data class TransactionState<out T : ContractState>(val contractState: T, val information: TransactionStateInformation) {

    /**
     * Creates a new instance of the [TransactionState] data class.
     *
     * @param contractState The [ContractState] component of the current [TransactionState] instance.
     * @param contract The class name of the contract associated with the transaction state.
     * @param notary The [PublicKey] of the notary associated with the transaction state.
     * @param encumbrance The index of an associated, encumbered state, or null if no encumbrance applies to the associated transaction state.
     * @param constraint The [CpkConstraint] associated with the transaction state.
     */
    constructor(contractState: T, contract: String, notary: PublicKey, encumbrance: Int?, constraint: CpkConstraint) :
            this(contractState, TransactionStateInformation(contract, notary, encumbrance, constraint))

    /**
     * Creates a new instance of the [TransactionState] data class.
     *
     * @param contractState The [ContractState] component of the current [TransactionState] instance.
     * @param notary The [PublicKey] of the notary associated with the transaction state.
     * @param encumbrance The index of an associated, encumbered state, or null if no encumbrance applies to the associated transaction state.
     * @param constraint The [CpkConstraint] associated with the transaction state.
     */
    constructor(contractState: T, notary: PublicKey, encumbrance: Int?, constraint: CpkConstraint) :
            this(contractState, contractState.getContractClassNameOrThrow(), notary, encumbrance, constraint)

    /**
     * Creates a new instance of the [TransactionState] data class.
     *
     * @param contractState The [ContractState] component of the current [TransactionState] instance.
     * @param notary The [PublicKey] of the notary associated with the transaction state.
     */
    constructor(contractState: T, notary: PublicKey) :
            this(contractState, notary, null, CpkConstraint.automaticPlaceholderCpkConstraint)

    /**
     * Creates a new instance of the [TransactionState] data class.
     *
     * @param contractState The [ContractState] component of the current [TransactionState] instance.
     * @param notary The [PublicKey] of the notary associated with the transaction state.
     * @param encumbrance The index of an associated, encumbered state, or null if no encumbrance applies to the associated transaction state.
     */
    constructor(contractState: T, notary: PublicKey, encumbrance: Int) :
            this(contractState, notary, encumbrance, CpkConstraint.automaticPlaceholderCpkConstraint)

    /**
     * Creates a new instance of the [TransactionState] data class.
     *
     * @param contractState The [ContractState] component of the current [TransactionState] instance.
     * @param notary The [PublicKey] of the notary associated with the transaction state.
     * @param constraint The [CpkConstraint] associated with the transaction state.
     */
    constructor(contractState: T, notary: PublicKey, constraint: CpkConstraint) :
            this(contractState, notary, null, constraint)

    /**
     * Creates a new instance of the [TransactionState] data class.
     *
     * @param contractState The [ContractState] component of the current [TransactionState] instance.
     * @param contract The class name of the contract associated with the transaction state.
     * @param notary The [PublicKey] of the notary associated with the transaction state.
     */
    constructor(contractState: T, contract: String, notary: PublicKey) :
            this(contractState, contract, notary, null, CpkConstraint.automaticPlaceholderCpkConstraint)

    /**
     * Creates a new instance of the [TransactionState] data class.
     *
     * @param contractState The [ContractState] component of the current [TransactionState] instance.
     * @param contract The class name of the contract associated with the transaction state.
     * @param notary The [PublicKey] of the notary associated with the transaction state.
     * @param encumbrance The index of an associated, encumbered state, or null if no encumbrance applies to the associated transaction state.
     */
    constructor(contractState: T, contract: String, notary: PublicKey, encumbrance: Int) :
            this(contractState, contract, notary, encumbrance, CpkConstraint.automaticPlaceholderCpkConstraint)

    /**
     * Creates a new instance of the [TransactionState] data class.
     *
     * @param contractState The [ContractState] component of the current [TransactionState] instance.
     * @param contract The class name of the contract associated with the transaction state.
     * @param notary The [PublicKey] of the notary associated with the transaction state.
     * @param constraint The [CpkConstraint] associated with the transaction state.
     */
    constructor(contractState: T, contract: String, notary: PublicKey, constraint: CpkConstraint) :
            this(contractState, contract, notary, null, constraint)
}

/**
 * Casts the current [TransactionState] to the specified type.
 *
 * @param T The underlying type of the [ContractState].
 * @param type The type of the [ContractState] to cast to.
 * @return Returns a new [TransactionState] instance cast to the specified type.
 * @throws IllegalArgumentException if the current [TransactionState] cannot be cast to the specified type.
 */
fun <T : ContractState> TransactionState<*>.cast(type: Class<T>): TransactionState<T> {
    return TransactionState(contractState.cast(type), information)
}

/**
 * Casts the current [TransactionState] to the specified type.
 *
 * @param T The underlying type of the [ContractState].
 * @return Returns a new [TransactionState] instance cast to the specified type.
 * @throws IllegalArgumentException if the current [TransactionState] cannot be cast to the specified type.
 */
inline fun <reified T : ContractState> TransactionState<*>.cast(): TransactionState<T> {
    return cast(T::class.java)
}
